src/k2/module/k23si/Module.cpp

```diff
@@ -350,14 +642,117 @@ K23SIPartitionModule::handleWrite(dto::K23SIWriteRequest&& request, dto::K23SI_M
+            request.designateTRH = false; // unset the flag and re-run
+            return handlePartialUpdate(std::move(request), std::move(sitMTR), deadline);
+        })
+        .handle_exception_type([this](TxnManager::ClientError&) {
+            // Failed to create
+            K2DEBUG("Partition: " << _partition << ", failed creating TR");
+            return RPCResponse(dto::K23SIStatus::AbortConflict("txn too old in write"), dto::K23SIPartialUpdateResponse{});
+        });
+    }
+
+    auto& versions = _indexer[request.key];
+    
+    if (!_validateStaleWrite(request, versions)) {
+        K2DEBUG("Partition: " << _partition << ", request too old for key " << request.key);
+        return RPCResponse(dto::K23SIStatus::AbortRequestTooOld("request too old in write"), dto::K23SIPartialUpdateResponse{});
+    }
+
+    // check to see if we should push or if we're coming after a push and the WI is still here
+    if (versions.size() > 0 && versions[0].status == dto::DataRecord::WriteIntent) {
+        auto& rec = versions[0];
+        auto& rqmtr = request.mtr;
+    
+        if (sitMTR == rec.txnId.mtr) {
+            K2DEBUG("Partition: " << _partition << ", post-push winner for key " << request.key);
+            // this is a post-PUSH request which won over the siting WI and we still have the WI in cache
+            _queueWICleanup(std::move(rec));
+            versions.pop_front();
+        }
+        else if (rec.txnId.mtr != rqmtr) {
+            // this is a update request finding a WI from a different transaction. Do another push with the remaining
+            // deadline time.
+            K2DEBUG("Partition: " << _partition << ", different WI found for key " << request.key);
+            sitMTR = rec.txnId.mtr;
+            return _doPush(request.collectionName, rec.txnId, request.mtr, deadline)
+                .then([this, sitMTR, request = std::move(request), deadline](auto&& winnerMTR) mutable {
+                    if (winnerMTR == sitMTR) {
+                        // sitting transaction won. Abort the incoming request
+                        K2DEBUG("Partition: " << _partition << ", push lost for key " << request.key);
+                        return RPCResponse(dto::K23SIStatus::AbortConflict("incumbent txn won in update push"), dto::K23SIPartialUpdateResponse{});
+                    }
+                    // incoming request won. re-run write logic
+                    K2DEBUG("Partition: " << _partition << ", push won for key " << request.key);
+                    return handlePartialUpdate(std::move(request), sitMTR, deadline);
+                });
+        }
+    }
+
+    // parse the partial record to full record
+    if ( !versions.size() ) {
+        // cannot parse partial record without a version
+        return RPCResponse(dto::K23SIStatus::KeyNotFound("can not partial update without a version"), dto::K23SIPartialUpdateResponse{});
+    }
+    if (!_parsePartialRecord(request, versions)) {
```

Comment:
You should check if versions[0] is a tombstone or not before calling this function. If it is a tombstone then you can return a KeyNotFound status.

Reply:
Check for tombstone is added and put together with the previous if() statement

```
if ( !versions.size() || versions[0].isTombstone) {
        // cannot parse partial record without a version
        return RPCResponse(dto::K23SIStatus::KeyNotFound("can not partial update with no/deleted version"), dto::K23SIPartialUpdateResponse{});
    }
```